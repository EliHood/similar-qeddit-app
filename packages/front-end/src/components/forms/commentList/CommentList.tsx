import React, { Fragment, useState, Ref } from 'react';
import Grid from '@material-ui/core/Grid';
import OurSecondaryButton from '../../../common/OurSecondaryButton';
import CommentListContainer from '../commentListContainer/commentListContainer';

function CommentList(props: any, ref: Ref<HTMLDivElement>) {
    const [showMore, setShowMore] = useState<number>(2);
    const [openModal, setOpenModal] = useState(false);
    const [showLessFlag, setShowLessFlag] = useState<boolean>(false);
    const the_comments = props.comments.length;
    const inc = showMore as any;
    const min = Math.min(2, the_comments - inc);
    const showComments = (e) => {
        e.preventDefault();

        if (inc + 2 && inc <= the_comments) {
            setShowMore(inc + 2);
            setShowLessFlag(true);
        } else {
            setShowMore(the_comments);
        }
    };

    const handleClickOpen = React.useCallback(() => {
        console.log('this got called owl man');
        setOpenModal(true);
    }, [setOpenModal]);
    const handleCloseModal = () => {
        setOpenModal(false);
    };

    const showLessComments = React.useCallback(
        (e) => {
            e.preventDefault();
            setShowMore(2);
            setShowLessFlag(false);
        },
        [setShowMore, setShowLessFlag],
    );
    const isBold = (comment) => (comment.userId === props.userId ? 800 : 400);

    // show comments by recent, and have the latest comment at the bottom, with the previous one just before it.
    const filterComments = props.comments
        .slice(0)
        .sort((a, b) => {
            const date1 = new Date(a.createdAt) as any;
            const date2 = new Date(b.createdAt) as any;

            return date2 - date1;
        })
        .slice(0, inc)
        .reverse();

    const showMoreComments = () => filterComments.map((comment, i) => (
        <div data-testid={`comment-show-more-${i}`} key={i}>
            <CommentListContainer ref={ref} comment={comment} openModal={openModal} handleCloseModal={handleCloseModal} isBold={isBold} handleClickOpen={handleClickOpen} {...props} />
        </div>
    ));

    const showMoreLess = (
        <div style={{ margin: '30px 0px' }} data-testid="show_more_less">
            {props.comments.length > 2 ? (
                <>
                    {min !== -1 && min !== -2 ? (
                        <>
                            {min !== 0 ? (
                                <OurSecondaryButton test="show_more_button" onClick={(e) => showComments(e)} component="span" color="secondary">
                                    View
                                    {' '}
                                    {min !== -1 && min !== -2 ? min : 0}
                                    {' '}
                                    More Comments
                                </OurSecondaryButton>
                            ) : (
                                <OurSecondaryButton test="show_less_button" onClick={(e) => showLessComments(e)} component="span" color="secondary">
                                    Show Less Comments
                                </OurSecondaryButton>
                            )}
                        </>
                    ) : (
                        <OurSecondaryButton test="show_less_button" onClick={(e) => showLessComments(e)} component="span" color="secondary">
                            Show Less Comments
                        </OurSecondaryButton>
                    )}
                </>
            ) : null}
        </div>
    );

    return (
        <Grid data-testid="comment-list-div">
            <>{showMoreLess}</>
            {showLessFlag === true ? (
            // will show most recent comments below
                showMoreComments()
            ) : (
                <>
                    {/* filter based on first comment  */}
                    {filterComments.map((comment, i) => (
                        <div key={i} ref={ref} className="comment">
                            <CommentListContainer ref={ref} comment={comment} openModal={openModal} handleCloseModal={handleCloseModal} isBold={isBold} handleClickOpen={handleClickOpen} {...props} />
                        </div>
                    ))}
                </>
            )}
        </Grid>
    );
}

export default React.forwardRef(CommentList) as React.RefForwardingComponent<HTMLDivElement, any>;
