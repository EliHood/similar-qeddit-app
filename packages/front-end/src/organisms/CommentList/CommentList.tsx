/* eslint-disable react/jsx-props-no-spreading */
import React, { forwardRef, useState, Ref } from 'react'
import Grid from '@material-ui/core/Grid'
import OurSecondaryButton from '../../molecules/OurSecondaryButton'
import CommentListContainer from '../CommentListContainer'

function CommentList(props: any, ref: Ref<HTMLDivElement>) {
    const { comments, userId, postId, opId } = props
    const [showMore, setShowMore] = useState<number>(2)
    const [openModal, setOpenModal] = useState(false)
    const [showLessFlag, setShowLessFlag] = useState<boolean>(false)
    const the_comments = comments.length
    const inc = showMore as any
    const min = Math.min(2, the_comments - inc)
    const showComments = (e) => {
        e.preventDefault()

        if (inc + 2 && inc <= the_comments) {
            setShowMore(inc + 2)
            setShowLessFlag(true)
        } else {
            setShowMore(the_comments)
        }
    }

    const handleClickOpen = React.useCallback(() => {
        setOpenModal(true)
    }, [setOpenModal])
    const handleCloseModal = () => {
        setOpenModal(false)
    }

    const showLessComments = React.useCallback(
        (e) => {
            e.preventDefault()
            setShowMore(2)
            setShowLessFlag(false)
        },
        [setShowMore, setShowLessFlag]
    )
    const isBold = (comment) => (comment.userId === userId ? 800 : 400)

    // show comments by recent, and have the latest comment at the bottom, with the previous one just before it.
    const filterComments = comments
        .slice(0)
        .sort((a, b) => {
            const date1 = new Date(a.createdAt) as any
            const date2 = new Date(b.createdAt) as any

            return date2 - date1
        })
        .slice(0, inc)
        .reverse()

    const showMoreComments = () =>
        filterComments.map((comment, i) => (
            <div data-testid={`comment-show-more-${i}`} key={comment.id}>
                <CommentListContainer
                    opId={opId}
                    ref={ref}
                    comment={comment}
                    openModal={openModal}
                    handleCloseModal={handleCloseModal}
                    isBold={isBold}
                    handleClickOpen={handleClickOpen}
                    postId={postId}
                />
            </div>
        ))

    const showMoreLess = (
        <div style={{ margin: '30px 0px' }} data-testid="show_more_less">
            {comments.length > 2 ? (
                <>
                    {min !== -1 && min !== -2 ? (
                        <>
                            {min !== 0 ? (
                                <OurSecondaryButton
                                    test="show_more_button"
                                    onClick={(e) => showComments(e)}
                                    component="span"
                                    color="secondary"
                                >
                                    View {min !== -1 && min !== -2 ? min : 0}{' '}
                                    More Comments
                                </OurSecondaryButton>
                            ) : (
                                <OurSecondaryButton
                                    test="show_less_button"
                                    onClick={(e) => showLessComments(e)}
                                    component="span"
                                    color="secondary"
                                >
                                    Show Less Comments
                                </OurSecondaryButton>
                            )}
                        </>
                    ) : (
                        <OurSecondaryButton
                            test="show_less_button"
                            onClick={(e) => showLessComments(e)}
                            component="span"
                            color="secondary"
                        >
                            Show Less Comments
                        </OurSecondaryButton>
                    )}
                </>
            ) : null}
        </div>
    )

    return (
        <Grid data-testid="comment-list-div">
            <>{showMoreLess}</>
            {showLessFlag === true ? (
                // will show most recent comments below
                showMoreComments()
            ) : (
                <>
                    {/* filter based on first comment  */}
                    {filterComments.map((comment, i) => (
                        <div key={comment.id} ref={ref} className="comment">
                            <CommentListContainer
                                opId={opId}
                                ref={ref}
                                comment={comment}
                                openModal={openModal}
                                handleCloseModal={handleCloseModal}
                                isBold={isBold}
                                handleClickOpen={handleClickOpen}
                                postId={postId}
                            />
                        </div>
                    ))}
                </>
            )}
        </Grid>
    )
}

export default forwardRef(CommentList);
