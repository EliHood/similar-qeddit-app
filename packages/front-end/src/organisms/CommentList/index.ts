import CommentList from './CommentList';

export default CommentList;