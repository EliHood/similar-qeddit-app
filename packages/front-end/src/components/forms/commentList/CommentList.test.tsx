import '@testing-library/jest-dom';
import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import CommentList from './CommentList';

const props = {
  user: {},
  postId: null,
  userId: null,
  currentUser: {},
  ref: {
    current: undefined,
  },
  comments: [
    {
      author: { username: 'barnowl', gravatar: 'https://api.adorable.io/avatars/400/bf1eed82fbe37add91cb4192e4d14de6.png', bio: null },
      comment_body: 'fsfsfsfsfs',
      createdAt: '2020-05-27T14:32:01.682Z',
      gifUrl: '',
      id: 520,
      postId: 28,
      updatedAt: '2020-05-27T14:32:01.682Z',
      userId: 9,
    },
    {
      author: { username: 'barnowl', gravatar: 'https://api.adorable.io/avatars/400/bf1eed82fbe37add91cb4192e4d14de6.png', bio: null },
      comment_body: 'fsfsfsfsfs',
      createdAt: '2020-05-27T14:32:01.682Z',
      gifUrl: '',
      id: 519,
      postId: 27,
      updatedAt: '2020-05-27T14:32:01.682Z',
      userId: 10,
    },
    {
      author: { username: 'barnowl2', gravatar: 'https://api.adorable.io/avatars/400/bf1eed82fbe37add91cb4192e4d14de6.png', bio: null },
      comment_body: 'fsfsfsfsfs',
      createdAt: '2020-05-27T14:32:01.682Z',
      gifUrl: '',
      id: 518,
      postId: 28,
      updatedAt: '2020-05-27T14:32:01.682Z',
      userId: 11,
    },
  ],
  deleteComment: jest.fn(),
};

describe('Should render <CommentList/>', () => {
  afterEach(cleanup);
  // your test

  it('should render <CommentList/>', () => {
    const commentList = render(<CommentList {...props} />);
    expect(commentList).toBeTruthy();
  });

  it('should render first comment', () => {
    const { getByTestId } = render(<CommentList {...props} />);
    const commentList = getByTestId('comment-list-div');
    expect(commentList.firstChild).toBeTruthy();
  });

  it('should render second child', () => {
    const { getByTestId } = render(<CommentList {...props} />);
    const commentList = getByTestId('comment-list-div');
    expect(commentList.lastChild).toBeTruthy();
  });

  it('should check comments', () => {
    const rtl = render(<CommentList {...props} />);

    expect(rtl.getByTestId('comment-list-div')).toBeTruthy();
    expect(rtl.getByTestId('comment-list-div')).toBeTruthy();

    expect(rtl.getByTestId('comment-list-div').querySelectorAll('.comment').length).toEqual(2);
  });
  it('should match snapshot', () => {
    const rtl = render(<CommentList {...props} />);
    expect(rtl).toMatchSnapshot();
  });

  it('should check more comments', () => {
    const { queryByTestId } = render(<CommentList {...props} />);
    const commentList = queryByTestId('comment-show-more');
    expect(commentList).toBeNull();
  });

  it('should fire show more action', () => {
    const { getByTestId, getAllByTestId } = render(<CommentList {...props} />);
    const showMore = getByTestId('show_more_button');
    fireEvent.click(showMore);
    expect(getAllByTestId(/comment-show-more/i).length).toEqual(3);
  });
});
