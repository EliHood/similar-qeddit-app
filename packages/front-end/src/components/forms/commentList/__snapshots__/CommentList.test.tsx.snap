// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render <CommentList/> should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root"
        data-testid="comment-list-div"
      >
        <div
          data-testid="show_more_less"
          style="margin: 30px 0px;"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
            data-testid="show_more_button"
            role="button"
            style="color: rgb(255, 255, 255); font-size: 11px;"
            tabindex="0"
            test="show_more_button"
          >
            <span
              class="MuiButton-label"
            >
              View 
              1
               More Comments
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
        </div>
        <div
          class="comment"
        >
          <ul
            class="MuiList-root MuiList-padding"
            data-testid="comment-list-container"
            style="padding-bottom: 20px;"
          >
            <img
              alt="gravatar"
              height="30"
              src="https://api.adorable.io/avatars/400/bf1eed82fbe37add91cb4192e4d14de6.png"
              style="margin: -10px 15px;"
              width="30"
            />
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft"
              style="display: inline-block; font-weight: 700; padding: 5px 0px;"
            >
              <span
                style="font-size: 12px; cursor: pointer; font-weight: 400;"
              >
                barnowl
              </span>
            </h6>
            <div
              data-testid="commentitem-wrapper"
              style="margin: 15px;"
            >
              <div
                class="MuiGrid-root MuiGrid-container"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-11"
                >
                  <div
                    class="markdownStyle"
                  >
                    <p>
                      fsfsfsfsfs
                    </p>
                  </div>
                  
                </div>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-alignLeft"
                style="font-size: 12px;"
              >
                Last Wednesday at 10:32 AM
              </p>
            </div>
          </ul>
        </div>
        <div
          class="comment"
        >
          <ul
            class="MuiList-root MuiList-padding"
            data-testid="comment-list-container"
            style="padding-bottom: 20px;"
          >
            <img
              alt="gravatar"
              height="30"
              src="https://api.adorable.io/avatars/400/bf1eed82fbe37add91cb4192e4d14de6.png"
              style="margin: -10px 15px;"
              width="30"
            />
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft"
              style="display: inline-block; font-weight: 700; padding: 5px 0px;"
            >
              <span
                style="font-size: 12px; cursor: pointer; font-weight: 400;"
              >
                barnowl
              </span>
            </h6>
            <div
              data-testid="commentitem-wrapper"
              style="margin: 15px;"
            >
              <div
                class="MuiGrid-root MuiGrid-container"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-11"
                >
                  <div
                    class="markdownStyle"
                  >
                    <p>
                      fsfsfsfsfs
                    </p>
                  </div>
                  
                </div>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-alignLeft"
                style="font-size: 12px;"
              >
                Last Wednesday at 10:32 AM
              </p>
            </div>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root"
      data-testid="comment-list-div"
    >
      <div
        data-testid="show_more_less"
        style="margin: 30px 0px;"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
          data-testid="show_more_button"
          role="button"
          style="color: rgb(255, 255, 255); font-size: 11px;"
          tabindex="0"
          test="show_more_button"
        >
          <span
            class="MuiButton-label"
          >
            View 
            1
             More Comments
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
      </div>
      <div
        class="comment"
      >
        <ul
          class="MuiList-root MuiList-padding"
          data-testid="comment-list-container"
          style="padding-bottom: 20px;"
        >
          <img
            alt="gravatar"
            height="30"
            src="https://api.adorable.io/avatars/400/bf1eed82fbe37add91cb4192e4d14de6.png"
            style="margin: -10px 15px;"
            width="30"
          />
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft"
            style="display: inline-block; font-weight: 700; padding: 5px 0px;"
          >
            <span
              style="font-size: 12px; cursor: pointer; font-weight: 400;"
            >
              barnowl
            </span>
          </h6>
          <div
            data-testid="commentitem-wrapper"
            style="margin: 15px;"
          >
            <div
              class="MuiGrid-root MuiGrid-container"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-11"
              >
                <div
                  class="markdownStyle"
                >
                  <p>
                    fsfsfsfsfs
                  </p>
                </div>
                
              </div>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-alignLeft"
              style="font-size: 12px;"
            >
              Last Wednesday at 10:32 AM
            </p>
          </div>
        </ul>
      </div>
      <div
        class="comment"
      >
        <ul
          class="MuiList-root MuiList-padding"
          data-testid="comment-list-container"
          style="padding-bottom: 20px;"
        >
          <img
            alt="gravatar"
            height="30"
            src="https://api.adorable.io/avatars/400/bf1eed82fbe37add91cb4192e4d14de6.png"
            style="margin: -10px 15px;"
            width="30"
          />
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft"
            style="display: inline-block; font-weight: 700; padding: 5px 0px;"
          >
            <span
              style="font-size: 12px; cursor: pointer; font-weight: 400;"
            >
              barnowl
            </span>
          </h6>
          <div
            data-testid="commentitem-wrapper"
            style="margin: 15px;"
          >
            <div
              class="MuiGrid-root MuiGrid-container"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-11"
              >
                <div
                  class="markdownStyle"
                >
                  <p>
                    fsfsfsfsfs
                  </p>
                </div>
                
              </div>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-alignLeft"
              style="font-size: 12px;"
            >
              Last Wednesday at 10:32 AM
            </p>
          </div>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
